src/TSHelper.ts

```diff
@@ -399,6 +345,80 @@ export class TSHelper {
             param => ts.isIdentifier(param.name) && param.name.originalKeywordKind === ts.SyntaxKind.ThisKeyword);
     }
 
+    public static findInClassOrAncestor(
+        classDeclaration: ts.ClassLikeDeclarationBase,
+        callback: (classDeclaration: ts.ClassLikeDeclarationBase) => boolean,
+        checker: ts.TypeChecker
+    ): ts.ClassLikeDeclarationBase
+    {
+        if (callback(classDeclaration)) {
+            return classDeclaration;
+        }
+
+        const extendsType = TSHelper.getExtendedType(classDeclaration, checker);
+        if (!extendsType) {
+            return undefined;
+        }
+
+        const symbol = extendsType.getSymbol();
+        const declaration = symbol.getDeclarations().find(ts.isClassLike);
+        if (!declaration) {
+            return undefined;
+        }
+
+        return TSHelper.findInClassOrAncestor(declaration, callback, checker);
+    }
+
+    public static hasSetAccessorInClassOrAncestor(
+        classDeclaration: ts.ClassLikeDeclarationBase,
+        checker: ts.TypeChecker
+    ): boolean
+    {
+        return TSHelper.findInClassOrAncestor(
+            classDeclaration,
+            c => c.members.some(ts.isSetAccessor),
+            checker
+        ) !== undefined;
+    }
+
+    public static getPropertyName(propertyName: ts.PropertyName): string | number | undefined {
+        if (ts.isIdentifier(propertyName) || ts.isStringLiteral(propertyName)) {
+            return propertyName.text;
+        } else if (ts.isNumericLiteral(propertyName)) {
+            return Number(propertyName.text);
```

Comment:
Is this line intentional? I would expect a function ***Name function to return strings

Reply:
It was intentional. I wanted to distinguish between string "0" and numerical 0 (the name was just being consistant with the ts.PropertyName type). But I just checked and it turns out TS treats those keys as being identical, so this is not necessary.